[APIv3] Specify status endpoint and edit form for WP

NobodysNightmare opened the pull request:
Specifies:
- [#17012](https://community.openproject.org/work_packages/17012)
- [#16971](https://community.openproject.org/work_packages/16971)

This PR introduces specification for:
- The full reading part of the statuses endpoint
- PATCH-ing the status of a work package
- edit forms to define which parts of a resource can be changed and which restrictions there are
## To Do
- [x] verify permissions needed to view statuses (currently "view work package anywhere")
- [x] Specify first set of action-links (will be required for forms)
- [x] Specify form endpoint
  - [x] basic structure
  - [x] abort errors (HTTP 4xx)
  - [x] indication of validation errors
## Open Questions
- [x] allowed values (links) vs. allowed values (non-links, e.g. strings)
  - values resource like `/foo?value=Cat`
- [x] handling of overly large lists of allowed values (e.g. users)
  - external collection resource (possibly paginated, possibly searchable)
- [x] what is the link of `changeParent`? Is it different from `update`?
  - same target, but answers question "can I change the parent"; answer can now be found in the form
## Remarks
- This PR also defines behaviour for PATCH/POST calls with completely wrong request bodies (i.e. non JSON; e.g. XML)...
  - specified as HTTP 400
  - most endpoints answer with HTTP 500
  - star/unstar require empty request body

NobodysNightmare committed:
specify status properties

NobodysNightmare committed:
specify basic readable status endpoint

NobodysNightmare committed:
move status to link section of work package

TeatroIO commented:
I've prepared a stage. <a href='http://feature-specify-status-endpoint.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

NobodysNightmare committed:
format WP-properties table prettier

NobodysNightmare committed:
specify 403 for statuses

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -787,6 +878,65 @@ Updates an activity's comment and, on success, returns the updated activity.
+
+    Returned if the client does not have sufficient permissions.
+    
+    **Required permission:** view workpackage (on any project)
```
spelling error "workpackage" -> "work package"

NobodysNightmare committed:
specify linked WP actions

only specify actions that:
- are linked in the implementation
- link into the api/v3-namespace (nothing outside the API)
- where link targets are already specified

+ also link the edit form that will be specified during this PR

NobodysNightmare committed:
start describing forms

NobodysNightmare committed:
fix aglio warning

[ci skip]

NobodysNightmare committed:
some explanation of field schema

field schema is quite limited as of now, but should suffice for current
needs

[ci skip]

NobodysNightmare committed:
add required attribute to field schema

[ci skip]

NobodysNightmare committed:
fix spelling error

[ci skip]

NobodysNightmare committed:
include an example

[ci skip]

NobodysNightmare committed:
fix more permission typos

[ci skip]

NobodysNightmare committed:
define error responses

includes HTTP 400 for all existing POST/PATCH endpoints

[ci skip]

NobodysNightmare committed:
formatting and stuff

[ci skip]

NobodysNightmare committed:
introduce error indication

[ci skip]

NobodysNightmare committed:
add guarantees about embedded properties

This allows clients to safely skip loading linked resources in case they
are already embedded.
Otherwise they could not safely assume which properties are embedded and
for which they'd need to resolve the link.

It should also simplify the API implementation as there is only
a single representation neccessary.

[ci skip]

NobodysNightmare committed:
remove changeParent action

does not seem to add any value beyond update/updateImmediately

[ci skip]

NobodysNightmare committed:
define payload for invalid/undefined values

[ci skip]

ulferts commented:
Notes from discussion:
- Renaming in form: "fields" -> "schema"
- Signal permissions in "fields"/"schema"? To avoid links like "change_parents"

NobodysNightmare committed:
introduce writable attribute

NobodysNightmare committed:
rename fields -> schema

[ci skip]

NobodysNightmare committed:
add soft specification of WP edit form

Avoid repetition in specification by only linking over to general
specs...

[ci skip]

NobodysNightmare commented:
I'd consider this spec ready for review.

Don't hesitate to point out what you don't like ^^

NobodysNightmare committed:
get rid of aglio-warning

[ci skip]

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -650,10 +889,22 @@ Updates an activity's comment and, on success, returns the updated activity.
 + Parameters
     + id (required, integer, `1`) ... Query id
-
+    
```
:scream:

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -389,6 +408,226 @@ Updates an activity's comment and, on success, returns the updated activity.
+
+It might happen that setting one property affects the allowed values for another property. Thus by changing a property A
+the current value of another property B might become invalid. If the client did not yet touch the value of B, the payload will
+contain a default value for that property. Nevertheless the client will also receive an apropriate validation error for value B.
```
Should we include here that the default value can also be empty?

NobodysNightmare replied:
If you mean that we will fill in "nothing" for properties where "nothing" is a valid value (e.g. assignee), I'd say this does not need to be mentioned.

If you mean anything else by saying "empty" (e.g. we will not provide anything), then I do not understand what you mean and we should talk ^^

**tl; dr:** I'd say no, but I have the feeling I have misunderstood you.

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -389,6 +408,226 @@ Updates an activity's comment and, on success, returns the updated activity.
+                            }
+                        }
+                    },
+                    "validationErrors": { }
```
Example validationError would be nice.

NobodysNightmare replied:
I had the fear that this would force me to omit the `commit` link... Lets see if I found a good approach to have both, an error and the commit-link in here.

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -389,6 +408,226 @@ Updates an activity's comment and, on success, returns the updated activity.
+Like `schema` the validation errors build a dictionary where the key is a property name.
+Each value is an error object that indicates the error that occured validating the corresponding property.
+
+There are only key value pairs for properties that failed validation, the element is empty if all validations succeeded.
```
It would help to point out again, that in the http sense, the existence of validationErrors still means an OK (200) response.

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -787,6 +1141,66 @@ Updates an activity's comment and, on success, returns the updated activity.
+            
+## View Status [GET]
+
+**NOT IMPLEMENTED**
```
While it might not be implemented, shouldn't we implement it in order to make the links to singular status as specified by this document valid?

NobodysNightmare replied:
I first thought that I'd mark everything with "not implemented" and the one implementing it will remove that mark.

But:
1. I did not mark everything
2. this might not be the best idea either...

I'll remove the mark for now

myabc commented on doc/apiv3-documentation.apib:
```diff
@@ -29,6 +30,10 @@ Read more: http://stateless.co/hal_specification.html
 
+HAL does not guarantee that embedded resources are embedded in their full representation, they might as well be
+partially represented (e.g. some properties can be left out).
+However in this API you have the guarantee that whenever a resource is **embedded**, it is embedded in its **full representation**.
```
:+1:

NobodysNightmare committed:
remove trailing whitespace

Hope @ulferts is happy now ;-)

[ci skip]

NobodysNightmare committed:
stress that a form is forgiving regarding errors

[ci skip]

NobodysNightmare committed:
remove not impl. markers from status

query keeps its marker, because this is indeed not implemented
and I don't see that somebody will do it soon.

[ci skip]

NobodysNightmare committed:
add an example validation error

[ci skip]

NobodysNightmare committed:
Merge branch 'dev' into feature/specify_status_endpoint

Conflicts:
	doc/apiv3-documentation.apib

myabc commented on doc/apiv3-documentation.apib:
```diff
@@ -427,7 +447,235 @@ Updates an activity's comment and, on success, returns the updated activity.
+* make writable properties of a resource discoverable
+* show to which values a property can be set
+* validate changes to a resource and indicate validation errors
+
```
@NobodysNightmare I know you already used the term _aid_ (which I like), but do we want to make it explicit that forms are optional?

NobodysNightmare replied:
Lets try: 650f246eff89f1a1fc0e5cf7896c114d54c18279

myabc commented on doc/apiv3-documentation.apib:
```diff
@@ -427,7 +447,235 @@ Updates an activity's comment and, on success, returns the updated activity.
+* validate changes to a resource and indicate validation errors
+
+A form is associated to a single resource and aids in performing changes on that resource.
+When posting to a form endpoint with an empty request body or an empty JSON object,
```
@NobodysNightmare there is no separate form endpoint? (or am I misunderstanding something?)

myabc replied:
(there definitely is a separate form endpoint, so this sentence is unclear to me)

NobodysNightmare replied:
We will (in the future) be using forms for more than only work-packages. This means that their documentation should be kept central, because the concept is always the same. However: The endpoint changes for each resource that is editable (`work_package/1/form`, `banana/42/form`, etc).

What you are reading here is the general explanation of forms, therefore I tried to use terms like "**a** form endpoint", to make clear that forms can be everywhere (usually "close" to their edited resource).

I could include an example: "a form endpoint (e.g. `/api/v3/work_packages/1/form`)", would that help?

myabc commented on doc/apiv3-documentation.apib:
```diff
@@ -427,7 +447,235 @@ Updates an activity's comment and, on success, returns the updated activity.
+                            }
+                        }
+                    },
+                    "validationErrors": {
```
@NobodysNightmare I don't understand the presence of `validationErrors` in the form.

NobodysNightmare replied:
Not in general or not in this particular example?

myabc replied:
@NobodysNightmare in general. ~~What would be the purpose?~~ (please ignore)

NobodysNightmare replied:
I wanted to start to link to [the purpose](https://github.com/opf/openproject/blob/2839c3f8c04a0f1838db88337506ac7b9f744e06/doc/apiv3-documentation.apib#L534-L539), but I am happy to ignore this part of the question ^^

In general: We want to indicate if the client did something wrong, i.e. that does not validate.
I see two approaches here:
- embedding those errors in the form (like proposed)
- send out a non-OK answer (e.g. HTTP 422)
  - In this case the client would receive the errors without a wrapping form

However this raises the question: How do you indicate which values are allowed in a hypothetical resource state.
Think of a resource with `A = 1, B = 2` allowed values for B are:

```
1, 2, 3 if A = 1
3, 4, 5 if A > 1
```

If you now set `A = 2`, then `B = 2` is not valid anymore: `urn:openproject-org:api:v3:errors:PropertyValueNotAvailableAnymore`

If the client now only receives an error object and no form that contains the link to the newly available values for `B`, the client will not be able to recover.

**Another perspective** could be the following:
If you always send the form back the client does not have to care about too much state, because the state is encoded in the form. If the client would receive a bunch of errors instead of a form, the client would have to go back to the last _good_ form it received and try again from that point on. So instead of going forward (always continue with the next state of the form), the client is forced to go back and forth on each error.

Is this the answer you were looking for or am I still missing the question? (I am not sure ^^)

myabc commented:
This is a really excellent first stab at forms!

NobodysNightmare committed:
stress the optional nature of forms

[ci skip]
